server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -192,6 +197,69 @@ public Weight createWeight(Query query, ScoreMode scoreMode, float boost) throws
         }
     }
 
+    @Override
+    protected LeafSlice[] slices(List<LeafReaderContext> leaves) {
+        return computeSlices(leaves, queueSizeBasedExecutor.threadPoolExecutor.getPoolSize(), MINIMUM_DOCS_PER_SLICE);
+    }
+
+    /**
+     * Each computed slice contains at least 10% of the total data in the leaves with a
+     * minimum given by the <code>minDocsPerSlice</code> parameter and the final number
+     * of {@link LeafSlice} will be equal or lower than the number of available threads.
+     */
+    /* pkg private for testing */
+    static LeafSlice[] computeSlices(List<LeafReaderContext> leaves, int numThreads, int minDocsPerSlice) {
+        // total number of documents to be searched
+        final int numDocs = leaves.stream().mapToInt(l -> l.reader().maxDoc()).sum();
+        // percentage of documents per slice, minumum 10%
+        final double percentageDocsPerThread = Math.max(MINIMUM_DOCS_PRECENT_PER_SLICE, 1.0 / numThreads);
```

Comment:
can you add an assert that `numThreads` is never 0? I think we should actually throw an exception if its less than `1` to be honest. and also have a test for it

Reply:
in other words pls. fix the division by zero potential here

Reply:
Added an explicit check and added to the tests.